spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/actions/SparkParquetFileMergeRunner.java

```diff
@@ -0,0 +1,215 @@
+          "Merging {} Parquet files using row-group level merge (group: {})",
+          group.rewrittenFiles().size(),
+          groupId);
+      mergeParquetFilesDistributed(groupId, group, schema);
+    } catch (Exception e) {
+      LOG.info(
+          "Row-group merge failed for group {}, falling back to standard Spark rewrite: {}",
+          groupId,
+          e.getMessage(),
+          e);
+      // Fallback to standard rewrite
+      super.doRewrite(groupId, group);
+    }
+  }
+
+  /**
+   * Validates if a group can be merged and returns the Parquet schema if successful.
+   *
+   * <p>This method checks all requirements for row-group merging and returns the schema if all
+   * checks pass. The returned schema can be reused to avoid redundant file reads.
+   *
+   * @param group the file group to validate
+   * @return MessageType schema if files can be merged, null otherwise
+   */
+  private MessageType canMergeAndGetSchema(RewriteFileGroup group) {
+    // Check if group expects exactly one output file
+    if (group.expectedOutputFiles() != 1) {
+      return null;
+    }
+
+    // Check if table has a sort order
+    if (table().sortOrder().isSorted()) {
+      return null;
+    }
+
+    // Check for delete files
+    boolean hasDeletes = group.fileScanTasks().stream().anyMatch(task -> !task.deletes().isEmpty());
+    if (hasDeletes) {
+      return null;
+    }
+
+    // Validate Parquet-specific requirements and get schema
+    return ParquetFileMerger.canMergeAndGetSchema(
+        Lists.newArrayList(group.rewrittenFiles()), table().io(), group.maxOutputFileSize());
+  }
+
+  /**
+   * Merges all input files in a group into a single output file.
+   *
+   * <p>This method assumes the group has been validated by {@link
+   * #canMergeAndGetSchema(RewriteFileGroup)} to have exactly one expected output file.
+   *
+   * @param schema the Parquet schema obtained from validation, reused to avoid redundant file reads
+   */
+  private void mergeParquetFilesDistributed(
+      String groupId, RewriteFileGroup group, MessageType schema) {
+    LOG.info(
+        "Merging {} Parquet files into 1 output file (group: {})",
+        group.rewrittenFiles().size(),
+        groupId);
+
+    // Get parameters needed for merge
+    long rowGroupSize =
+        PropertyUtil.propertyAsLong(
+            table().properties(),
+            TableProperties.PARQUET_ROW_GROUP_SIZE_BYTES,
+            TableProperties.PARQUET_ROW_GROUP_SIZE_BYTES_DEFAULT);
+    FileIO fileIO = table().io();
+    Table serializableTable = table();
+
+    // Execute merge on an executor
+    // Use parallelize to run the merge task on an executor
+    JavaSparkContext jsc = JavaSparkContext.fromSparkContext(spark().sparkContext());
+    JavaRDD<RewriteFileGroup> taskRDD = jsc.parallelize(Lists.newArrayList(group), 1);
+    DataFile newFile =
+        taskRDD
+            .map(task -> mergeFilesForTask(task, rowGroupSize, fileIO, serializableTable, schema))
```

Comment:
Using lambda while sending the task information can be risky due to shading. 
We are already passing Parquet related parameters like schema which are shaded in Iceberg.
I would suggest using an inner class rather than lambda to avoid executor serialization issues.

cc @pvary